File: 42.trapping-rain-water.cpp
/*
 * @lc app=leetcode id=42 lang=cpp
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
class Solution {
   public:
    int trap(vector<int>& height) {
        int ans = 0, l = 0, r = height.size() - 1, level = 0;
        while (l < r) {
            int lower = (height[l] > height[r] ? height[r--] : height[l++]);
            level = max(lower, level);
            ans += level - lower;
        }
        return ans;
    }
};
// @lc code=end

File: 42.Trapping-Rain-Water.ts
function trap(height: number[]): number {
  let res = 0;
  let left = 0;
  let right = height.length - 1;
  let level = 0;
  while (left < right) {
    let lower = height[left] > height[right] ? height[right--] : height[left++];
    level = Math.max(lower, level);
    res += level - lower;
  }
  return res;
};